### ISSUE TITLE
Signals and exec() interaction

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#12352]
Migrated from [rt.perl.org#114586](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=114586) (status was 'open')

Searchable as RT114586$

**Comment by p5pRT** [comment]
### From @dgl

In despatch\_signals in mg\.c there's a comment explaining how perl
tries to emulate a behaviour of real signals in safe signals&#8203;:

&nbsp;            /\* From sigaction\(2\) \(FreeBSD man page\)&#8203;:
&nbsp;             \* | Signal routines normally execute with the signal that
&nbsp;             \* | caused their invocation blocked\, but other signals may
&nbsp;             \* | yet occur\.
&nbsp;             \* Emulation of this behavior \(from within Perl\) is enabled
&nbsp;             \* using sigprocmask
&nbsp;             \*/

However there's an issue in the emulation&#8203;: if a signal handler
directly calls exec\(\) the signal is never unblocked\, because the
destructor/clean up function set up in Perl\_despatch\_signals isn't
called\.

David

**Comment by p5pRT** [comment]
### From @Leont

On Sat\, Aug 25\, 2012 at 1&#8203;:17 PM\, David Leadbeater
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> \# New Ticket Created by  David Leadbeater
> \# Please include the string&#8203;:  \[perl \#114586\]
> \# in the subject line of all future correspondence about this issue\.
> \# \<URL&#8203;: https://rt-archive.perl.org/perl5/Ticket/Display.html?id=114586 >
>
>
> In despatch\_signals in mg\.c there's a comment explaining how perl
> tries to emulate a behaviour of real signals in safe signals&#8203;:
>
>             /\* From sigaction\(2\) \(FreeBSD man page\)&#8203;:
>              \* | Signal routines normally execute with the signal that
>              \* | caused their invocation blocked\, but other signals may
>              \* | yet occur\.
>              \* Emulation of this behavior \(from within Perl\) is enabled
>              \* using sigprocmask
>              \*/
>
> However there's an issue in the emulation&#8203;: if a signal handler
> directly calls exec\(\) the signal is never unblocked\, because the
> destructor/clean up function set up in Perl\_despatch\_signals isn't
> called\.

But is that really different from the unsafe version? I'd expect it
not to be\, though I don't think POSIX is explicit about this case\.

Leon

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'
